pkg/engine/jmespath/functions.go

```diff
@@ -1061,56 +1061,88 @@ func jpRandom(arguments []interface{}) (interface{}, error) {
-	}
-
-	cert, err := x509.ParseCertificate(p.Bytes)
-	if err != nil {
-		return res, err
-	}
-
-	buf := new(bytes.Buffer)
-	if fmt.Sprint(cert.PublicKeyAlgorithm) == "RSA" {
-		spki := cryptobyte.String(cert.RawSubjectPublicKeyInfo)
-		if !spki.ReadASN1(&spki, cryptobyte_asn1.SEQUENCE) {
-			return res, errors.New("writing asn.1 element to 'spki' failed")
-		}
-		var pkAISeq cryptobyte.String
-		if !spki.ReadASN1(&pkAISeq, cryptobyte_asn1.SEQUENCE) {
-			return res, errors.New("writing asn.1 element to 'pkAISeq' failed")
-		}
-		var spk asn1.BitString
-		if !spki.ReadASN1BitString(&spk) {
-			return res, errors.New("writing asn.1 bit string to 'spk' failed")
+	} else if block, _ := pem.Decode([]byte(input.String())); block == nil {
+		return nil, errors.New("failed to decode PEM block")
+	} else {
+		buf := new(bytes.Buffer)
+		switch block.Type {
+		case "CERTIFICATE":
+			var cert *x509.Certificate
+			if cert, err = x509.ParseCertificate(block.Bytes); err != nil {
+				return nil, err
+			}
+			if fmt.Sprint(cert.PublicKeyAlgorithm) != "RSA" {
+				return nil, errors.New("certificate should use rsa algorithm")
+			}
+			spki := cryptobyte.String(cert.RawSubjectPublicKeyInfo)
+			N, E, err := parseCryptobyte(&spki)
+			if err != nil {
+				return res, err
+			}
+
+			cert.PublicKey = PublicKey{
+				N: N,
+				E: E,
+			}
+
+			enc := json.NewEncoder(buf)
+			err = enc.Encode(cert)
+			if err != nil {
+				return res, err
+			}
+		case "CERTIFICATE REQUEST":
+			var cr *x509.CertificateRequest
+			if cr, err = x509.ParseCertificateRequest(block.Bytes); err != nil {
+				return nil, err
+			}
+			if fmt.Sprint(cr.PublicKeyAlgorithm) != "RSA" {
+				return nil, errors.New("certificate eequest should use rsa algorithm")
+			}
+			spki := cryptobyte.String(cr.RawSubjectPublicKeyInfo)
+			N, E, err := parseCryptobyte(&spki)
+			if err != nil {
+				return res, err
+			}
+			cr.PublicKey = PublicKey{
+				N: N,
+				E: E,
+			}
+
+			enc := json.NewEncoder(buf)
+			err = enc.Encode(cr)
+			if err != nil {
+				return res, err
+			}
+		default:
+			return nil, errors.New("PEM block neither contains a CERTIFICATE or CERTIFICATE REQUEST")
 		}
-		kk, err := x509.ParsePKCS1PublicKey(spk.Bytes)
-		if err != nil {
+		if err := json.Unmarshal(buf.Bytes(), &res); err != nil {
 			return res, err
 		}
 
-		cert.PublicKey = PublicKey{
-			N: kk.N.String(),
-			E: kk.E,
-		}
-
-		enc := json.NewEncoder(buf)
-		err = enc.Encode(cert)
-		if err != nil {
-			return res, err
-		}
+		return res, nil
 	}
+}
 
-	if err := json.Unmarshal(buf.Bytes(), &res); err != nil {
-		return res, err
+func parseCryptobyte(spki *cryptobyte.String) (string, int, error) {
+	if !spki.ReadASN1(spki, cryptobyte_asn1.SEQUENCE) {
+		return "", -1, errors.New("writing asn.1 element to 'spki' failed")
 	}
-
-	return res, nil
+	var pkAISeq cryptobyte.String
+	if !spki.ReadASN1(&pkAISeq, cryptobyte_asn1.SEQUENCE) {
+		return "", -1, errors.New("writing asn.1 element to 'pkAISeq' failed")
+	}
```

Comment:
Do we need this ? What are we doing with `pkAISeq` ?

Reply:
These checks were in place in the function that I built on top of

Reply:
Actually, it was there already in the x509 function when I was adding csr support and, `spki.ReadASN1BitString(&spk)` fails if we remove that part.